HPCC-27692 Add histogram metric type

kenrowland opened the pull request:
Added histogram metric

Signed-Off-By: Ken Rowland kenneth.rowland@lexisnexisrisk.com

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Cloud-compatibility
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

Commit:
HPCC-27692 Add historgram metric type

Added histogram metric

Signed-Off-By: Ken Rowland kenneth.rowland@lexisnexisrisk.com

kenrowland commented:
@ghalliday First pass at a histogram metric. It follows the requirements for a Prometheus histogram with the exception that it does not accumulate measurements as the bucket limits increase. That can be done by the sink as a metrics storage dependency. 

The class is templated to allow for different types for bucket limits.

ghalliday commented on system/jlib/jmetrics.hpp:
```diff
@@ -237,6 +236,101 @@ class CustomMetric : public MetricBase
 };
 
 
+template <typename T>
```
I don't think the template parameter is a good idea for a few reasons
* It complicates all the types, and I'm not sure it adds much
* The sinks need to be able to get at the results - they need a virtual function in IMetric that they can call to get them, and that is not possible if the type is a template.

ghalliday commented on system/jlib/jmetrics.hpp:
```diff
@@ -237,6 +236,101 @@ class CustomMetric : public MetricBase
+        {
+            for (auto &b: bucketDefs)
+            {
+                buckets.emplace_back(Bucket<T>{b.label, b.limit});
```
If you're using emplace_back() I think you should be able to say

        buckets.emplace_back(b.label, b.limit);

ghalliday replied:
Have you tried again now it is no longer a template - it compiles for me.

ghalliday commented on system/jlib/jmetrics.hpp:
```diff
@@ -237,6 +236,101 @@ class CustomMetric : public MetricBase
+        values.sum = sum;
+        for (auto const &bucket: buckets)
+        {
+            values.bucketValues.emplace_back(BucketValue{bucket.label, bucket.count});
```
minor:    emplace_back(bucket.label, bucket.count);

ghalliday commented on system/jlib/jmetrics.hpp:
```diff
@@ -237,6 +236,101 @@ class CustomMetric : public MetricBase
+protected:
+    Bucket<T> &getBucket(T measurement)
+    {
+        for (auto b = buckets.begin(); b != buckets.end(); b++)
```
minor: use `for (auto &bucket: buckets)` as above

ghalliday commented on testing/unittests/metrics/MetricFrameworkTests.cpp:
```diff
@@ -308,6 +309,62 @@ class MetricFrameworkTests : public CppUnit::TestFixture
+
+        //
+        // Verify that the correct number of buckets is returned
+        values = pHistogram->queryValues();
```
To mimic use from a sink this test needs to use an IMetric to retrieve the values.

ghalliday replied:
Yes, the sink will need to know it is a histogram.  I don't have a strong opinion on the interface, but the minimum would be something like

__uint64 queryValue();// returning the sum of the values and
std::vector<__uint64> getHistogramValues(); // returning the counts.  (The total count could be returned by a separate function, or caulcated by summing all these values.)

ghalliday commented on system/jlib/jmetrics.hpp:
```diff
@@ -237,6 +236,101 @@ class CustomMetric : public MetricBase
+        values.sum = sum;
+        for (auto const &bucket: buckets)
+        {
+            values.bucketValues.emplace_back(BucketValue{bucket.label, bucket.count});
```
As far as I can see the Prometheus standard is for the counts to accumulate.  i.e. if the counts are {1 ,2, 2, 1} they would be reported as { 1, 3, 5, 6 }.  Is this the responsibility of the sink or the histogram?  I don't think it matters either way, but it needs to be clearly documented.

kenrowland replied:
I opted for each bucket maintained by the metric to be the count of measurements for that bucket only. So, yes it's the responsibility of the sink to convert to the representation required. Additionally, the accumulation only needs to be done at report time reducing the overhead of recording a measurement. As far as documenting, I added a simple explanation in the metrics design doc as part of this PR.

ghalliday commented on system/jlib/jmetrics.hpp:
```diff
@@ -237,6 +236,101 @@ class CustomMetric : public MetricBase
+    HistogramMetricValues<T> queryValues()
+    {
+        HistogramMetricValues<T> values;
+        values.count = count;
```
This could be inconsistent, does it matter? It is possible the count could be higher than the sum of the bucket counts if other calls to recordMeasurement() come in while it is executing.
It may be worth considering if a critical section is better than atomics.  On my machine uncontended critical sections have an overhead of ~16ns, atomic increments have an overhead of ~7ns. (Run `unittests AtomicTimingTest` to see full details.)

ghalliday commented on system/jlib/jmetrics.hpp:
```diff
@@ -237,6 +236,101 @@ class CustomMetric : public MetricBase
+
+    HistogramMetricValues<T> queryValues()
+    {
+        HistogramMetricValues<T> values;
```
worth calling

    values.bucketsValues.reserve(buckets.size()+1);

to avoid reallocation of the result.

ghalliday commented on system/jlib/jmetrics.hpp:
```diff
@@ -237,6 +236,101 @@ class CustomMetric : public MetricBase
+        values.sum = sum;
+        for (auto const &bucket: buckets)
+        {
+            values.bucketValues.emplace_back(BucketValue{bucket.label, bucket.count});
```
Does the result need to include the label?  This should be available to the sink from the meta.

ghalliday replied:
Sorry for the confusion, I had forgotten what the meta data was used for.

Reporting the metrics isn't called very often, so the efficiency is not that important, but cloning the string labels each time is less efficient.

ghalliday commented on testing/unittests/metrics/MetricFrameworkTests.cpp:
```diff
@@ -308,6 +309,62 @@ class MetricFrameworkTests : public CppUnit::TestFixture
+        CPPUNIT_ASSERT_EQUAL(static_cast<int>(values.sum), sumMeasurements);
+        CPPUNIT_ASSERT_EQUAL(static_cast<int>(values.count), countMeasurements);
+
+        //
```
I doubt there is a problem, but the unit test should also test increment other buckets.

Potentially worth adding a 

checkResults(values, sum, count, { 0, 2, 0, 0});

 to reduce the code checking the test cases.

kenrowland replied:
Added a method to simplify testing the entire results.

ghalliday commented on system/jlib/jmetrics.hpp:
```diff
@@ -237,6 +236,101 @@ class CustomMetric : public MetricBase
+class jlib_decl HistogramMetric : public MetricBase
+{
+public:
+    HistogramMetric(const char *name, const char *desc, StatisticMeasure _units, const std::vector<BucketInfo<T>> &bucketDefs, const MetricMetaData &_metaData = MetricMetaData()) :
```
structure: bucket defs should be part of the metric meta data (and no default parameter)

kenrowland replied:
There is a comment later that explains why I made the definitions separate from meta data.

ghalliday reviewed (commented):
@kenrowland it looks broadly right.  A few different comments.

ghalliday commented on devdoc/Metrics.rst:
```diff
@@ -290,6 +290,19 @@ metric value as the custom metric class provides no update methods. The componen
+maximum bucket limit. This is sometimes known as the "inf" bucket.
+
+Some storage systems, such as Prometheus, require each bucket to accumulate its measurements with
+the previous bucket(s). It is the repsonsbility of the sink to accumulate values as needed.
```
typo responsibility

kenrowland commented:
@ghalliday 
Addressed comments. I opted for three methods to return values from the metric:
queryValue - returns the total of all measurements which is inline with what other metrics do
querySum - Total sum of counts (all bucket counts)
queryHistogramValues - individual bucket counts (including inf)

The caller could sum the counts if needed instead of calling querySum

I also separated labels from the histogram values. There is a queryLabels method now. The intent is the sink would call this only  once to prefill a structure used to report values. This way the labels are not constantly recreated.

Version of my compiler still doesn't like not specifying the structure type in the emplace_back. In the end it's more readable anyway.

ghalliday commented on testing/unittests/metrics/MetricFrameworkTests.cpp:
```diff
@@ -308,6 +309,65 @@ class MetricFrameworkTests : public CppUnit::TestFixture
+    }
+
+
+    void checkHistogramBucketResult(std::shared_ptr<HistogramMetric> pHistogram, int expectedSum, int expectedCount, const std::vector<int> &expectedValues)
```
This function needs to take an IMetric as a parameter - so it can match the logic that the sinks need to execute.  (The functions are not currently accessible to the sinks...)

kenrowland replied:
Converted parameter to IMetric as requested. Add dynamic pointer cast to histogram metric type to gain access to histogram specific methods. 

It is expected that a sink will check the type of each metric in it's report list and for any encountered histograms, execute a dynamic pointer cast to convert the pointer type and gain access to the necessary methods.

ghalliday reviewed (changes requested):
This isn't what I expected - it is not well encapsulated.

The sinks should only call functions in the IMetric interface - they should know nothing about the implementation classes.  (Ideally the implementations would be in separate header files that are not even included by the sink code.)

Also, please do not squash commits until the PR has been approved/someone requests you squash it.

kenrowland commented:
Added histogram methods to metric interface. Base class implementation returns empty values.

ghalliday commented on system/jlib/jmetrics.hpp:
```diff
@@ -237,6 +249,56 @@ class CustomMetric : public MetricBase
+};
+
+
+struct BucketValue
```
This can now be removed.

ghalliday commented on system/jlib/jmetrics.hpp:
```diff
@@ -110,8 +120,10 @@ class jlib_decl MetricBase : public IMetric
-    const MetricMetaData &queryMetaData() const { return metaData; }
+    const MetricMetaData &queryMetaData() const override { return metaData; }
     StatisticMeasure queryUnits() const override { return units; }
+    virtual std::vector<__uint64> queryHistogramValues() const { return {}; }
```
trivial: can also be override

ghalliday commented on system/jlib/jmetrics.hpp:
```diff
@@ -237,6 +249,56 @@ class CustomMetric : public MetricBase
+    }
+
+    void recordMeasurement(__uint64 measurement);
+    std::vector<__uint64> queryHistogramValues() const;
```
minor: virtual and override for clarity (and next function)

ghalliday reviewed (commented):
@kenrowland looks good.  A few final minor comments.

kenrowland commented:
Updated based on recent comments.

ghalliday commented on testing/unittests/metrics/MetricFrameworkTests.cpp:
```diff
@@ -308,6 +309,62 @@ class MetricFrameworkTests : public CppUnit::TestFixture
+
+    void checkHistogramBucketResult(std::shared_ptr<IMetric> pMetric, int expectedSum, const std::vector<int> &expectedValues)
+    {
+        std::shared_ptr<HistogramMetric> pHistogram;
```
minor: These lines should be deleted, and the pMetric used - so this code will mirror the calls the sink will perform.

kenrowland replied:
Removed cast to histogram pointer type in favor of IMetric interface use only.

ghalliday reviewed (commented):
@kenrowland looks good.  One last bit of tidying up on the test case to ensure the interface is correct.

ghalliday commented:
@kenrowland merged, but then noticed use had the wrong issue number in your commit message.  Should have been HPCC-27962.  Will have to stand as-is.
